### PR TITLE
NIM on GKE Tutorial

### PR DESCRIPTION
**Pull request by brandonroyal**
Tutorial for running NIM on GKE using A100 GPUs

### CODE REVIEW

**Comment by brandonroyal** [comment]
@imreddy13 @roberthbailey Can you pls review?

File: tutorials-and-examples/nvidia-nim/README.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+## Prerequisites
+### Prerequisites
+
+> [!IMPORTANT]
+> Before you proceed further, ensure you have the NVIDIA AI Enterprise License (NVAIE) to access the NIMs.  To get started, go to [build.nvidia.com](https://build.nvidia.com/explore/discover?signin=true) and provide your company email address
+
+* [Google Cloud Project](https://console.cloud.google.com) with billing enabled
+* [gcloud CLI](https://cloud.google.com/sdk/docs/install)
+* [gcloud kubectl](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl#install_kubectl)
+<!-- * [Terraform](https://developer.hashicorp.com/terraform/tutorials/gcp-get-started/install-cli) -->
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: Fixed

File: tutorials-and-examples/nvidia-nim/README.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
+# NVIDIA NIM on GKE with NVIDIA A100s
+
+## Prerequisites
+### Prerequisites
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: Good call.  Added to the prerequisites.

File: tutorials-and-examples/nvidia-nim/README.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+> [!IMPORTANT]
+> Before you proceed further, ensure you have the NVIDIA AI Enterprise License (NVAIE) to access the NIMs.  To get started, go to [build.nvidia.com](https://build.nvidia.com/explore/discover?signin=true) and provide your company email address
+
+* [Google Cloud Project](https://console.cloud.google.com) with billing enabled
+* [gcloud CLI](https://cloud.google.com/sdk/docs/install)
+* [gcloud kubectl](https://cloud.google.com/kubernetes-engine/docs/how-to/cluster-access-for-kubectl#install_kubectl)
+<!-- * [Terraform](https://developer.hashicorp.com/terraform/tutorials/gcp-get-started/install-cli) -->
+* [Git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
+*  [yq](https://pypi.org/project/yq/)
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: Fixed.  You're correct, this should be `jq`

File: tutorials-and-examples/nvidia-nim/README.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+3. Create a nodepool
+```bash
+gcloud container node-pools create ${MACH?}-node-pool --cluster nim-demo \
+   --accelerator type=${GPU_TYPE?},count=${GPU_COUNT?},gpu-driver-version=latest \
+  --machine-type ${MACH?} \
+  --ephemeral-storage-local-ssd=count=${GPU_COUNT?} \
+  --enable-autoscaling --enable-image-streaming \
+  --num-nodes=1 --min-nodes=1 --max-nodes=3 \
+  --node-locations ${REGION?}-b \
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: fixed.

File: tutorials-and-examples/nvidia-nim/README.md [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+1. Clone the nim-deploy repository
+```bash
+git clone https://github.com/NVIDIA/nim-deploy.git
+cd nim-deploy/helm
+```
+
+2. Deploy chart with minimal configurations
+```bash
+helm --namespace nim install demo-nim nim-llm/ --set model.ngcAPIKey=$NGC_CLI_API_KEY --set persistence.enabled=true --set persistence.existingClaim=model-store-pvc
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md:125]: Good catch.  Looks like the secrets are redundant.  Fixed

File: tutorials-and-examples/nvidia-nim/README.md [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+## Test the NIM
+1. Expose the service
+```bash
+kubectl port-forward services/demo-nim-nim-llm 8000
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: Fixed.

File: tutorials-and-examples/nvidia-nim/.gitignore [hunk 7]
```diff
@@ -0,0 +1,4 @@
+jupyter-server.yaml
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/.gitignore]: Deleted the entire .gitignore file as it's not needed.

File: tutorials-and-examples/nvidia-nim/README.md [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+3. Ensure you have access to the repository by listing the models
+```bash
+ngc registry model list
+```
+
+4. Create a Kuberntes namespace and switch context to that namespace
+```bash
+kubectl create namespace nim
+kubectl config set-context --current --namespace nim
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: Agreed.  Fixed.

File: tutorials-and-examples/nvidia-nim/README.md [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+```
+
+5. Create Kubernetes secrets to enable access to NGC resources from within your cluster
+```bash
+kubectl -n nim create secret docker-registry registry-secret --docker-server=nvcr.io --docker-username='$oauthtoken' --docker-password=$NGC_CLI_API_KEY
+kubectl -n nim create secret generic ngc-api --from-literal=NGC_CLI_API_KEY=$NGC_CLI_API_KEY
+```
+
+## Deploy a PVC to persist the model
+1. Clone this repository
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: Agreed. Fixed.

File: tutorials-and-examples/nvidia-nim/README.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+2. Create a PVC to persist the model weights - recommended for deployments with more than one (1) replica.  Save the following yaml as `pvc.yaml` or use existing file in this repository
```
[the remark this answers is not in the document]
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: Good call.  Fixed.

File: tutorials-and-examples/nvidia-nim/README.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+```bash
+ngc config set
+```
+
+3. Ensure you have access to the repository by listing the models
+```bash
+ngc registry model list
+```
+
+4. Create a Kuberntes namespace and switch context to that namespace
```
Comment (roberthbailey) [inline tutorials-and-examples/nvidia-nim/README.md:85]: nit: Remove "and switch context to that namespace"

File: tutorials-and-examples/nvidia-nim/pvc.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2024 Google LLC
```
Comment (roberthbailey) [inline tutorials-and-examples/nvidia-nim/pvc.yaml:1]: Please remove this file from the pull request.

File: tutorials-and-examples/nvidia-nim/README.md [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+```bash
+kubectl create namespace nim
+```
+
+## Deploy a PVC to persist the model
+1. Create a PVC to persist the model weights - recommended for deployments with more than one (1) replica.  Save the following yaml as `pvc.yaml` or use existing file in this repository
```
Comment (roberthbailey) [inline tutorials-and-examples/nvidia-nim/README.md:91]: nit: Remove "or use existing file in this repository"

File: tutorials-and-examples/nvidia-nim/README.md (hunk 11 above)
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: Fixed in the latest commit

File: tutorials-and-examples/nvidia-nim/pvc.yaml (hunk 12 above)
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/pvc.yaml]: Deleted.

File: tutorials-and-examples/nvidia-nim/README.md (hunk 13 above)
Comment (brandonroyal) [reply tutorials-and-examples/nvidia-nim/README.md]: Cleaned this up in an earlier commit.  Make sure you have the latest

**Review by roberthbailey — APPROVED** [review]

**Pull request merged**